src/dex/sundaeswap-v3.ts

```diff
@@ -0,0 +1,238 @@
+        const swapFee: bigint = ((receive * BigInt(Math.floor(liquidityPool.poolFeePercent * 100))) + BigInt(10000) - 1n) / 10000n;
+
+        return receive + swapFee;
+    }
+
+    estimatedReceive(liquidityPool: LiquidityPool, swapInToken: Token, swapInAmount: bigint): bigint {
+        const [reserveIn, reserveOut]: bigint[] = correspondingReserves(liquidityPool, swapInToken);
+
+        const swapFee: bigint = ((swapInAmount * BigInt(Math.floor(liquidityPool.poolFeePercent * 100))) + BigInt(10000) - 1n) / 10000n;
+
+        return reserveOut - (reserveIn * reserveOut) / (reserveIn + swapInAmount - swapFee);
+    }
+
+    priceImpactPercent(liquidityPool: LiquidityPool, swapInToken: Token, swapInAmount: bigint): number {
+        const reserveIn: bigint = tokensMatch(swapInToken, liquidityPool.assetA)
+            ? liquidityPool.reserveA
+            : liquidityPool.reserveB;
+
+        return (1 - (Number(reserveIn) / Number(reserveIn + swapInAmount))) * 100;
+    }
+
+    public async buildSwapOrder(liquidityPool: LiquidityPool, swapParameters: DatumParameters, spendUtxos: SpendUTxO[] = []): Promise<PayToAddress[]> {
+        // TODO
+        const scooperFee: SwapFee | undefined = this.swapOrderFees(liquidityPool).find((fee: SwapFee) => fee.id === 'scooperFee');
+        const deposit: SwapFee | undefined = this.swapOrderFees(liquidityPool).find((fee: SwapFee) => fee.id === 'deposit');
+
+        if (! scooperFee || ! deposit) {
+            return Promise.reject('Parameters for datum are not set.');
+        }
+
+        const swapInToken: string = (swapParameters.SwapInTokenPolicyId as string) + (swapParameters.SwapInTokenAssetName as string);
+        const swapOutToken: string = (swapParameters.SwapOutTokenPolicyId as string) + (swapParameters.SwapOutTokenAssetName as string);
+        const swapDirection: number = [swapInToken, swapOutToken].sort((a: string, b: string) => {
+            return a.localeCompare(b);
+        })[0] === swapInToken ? 0 : 1;
+
+        swapParameters = {
+            ...swapParameters,
+            [DatumParameterKey.ScooperFee]: scooperFee.value,
+            [DatumParameterKey.Action]: swapDirection,
+        };
+
+        const datumBuilder: DefinitionBuilder = new DefinitionBuilder();
+        await datumBuilder.loadDefinition(order)
+            .then((builder: DefinitionBuilder) => {
+                builder.pushParameters(swapParameters);
+            });
+
+        return [
+            this.buildSwapOrderPayment(
+                swapParameters,
+                {
+                    address: this.orderAddress,
+                    addressType: AddressType.Contract,
+                    assetBalances: [
+                        {
+                            asset: 'lovelace',
+                            quantity: scooperFee.value + deposit.value,
+                        },
+                    ],
+                    datum: datumBuilder.getCbor(),
+                    isInlineDatum: false,
```

Comment:
Yes; technically we support non-inline datums in the contracts in case there is ever a need for it, but some of the offchain infrastructure might choke on witness datums.

Reply:
Fixed this in 1cc0702ae1297095146d2300c48fec132f8dd993